libraries/GCS_MAVLink/GCS_Common.cpp

```diff
@@ -583,81 +705,163 @@ void GCS_MAVLINK::handle_mission_set_current(AP_Mission &mission, mavlink_messag
+
+    dest_sysid = msg.sysid;       // record system id of GCS who wants to upload the mission
+    dest_compid = msg.compid;     // record component id of GCS who wants to upload the mission
+
+    link = &_link;
+
+    timelast_request_ms = AP_HAL::millis();
+    link->send_message(next_item_ap_message_id());
+}
+
+void MissionItemProtocol::handle_mission_count(
+    GCS_MAVLINK &_link,
+    const mavlink_mission_count_t &packet,
+    const mavlink_message_t &msg)
+{
+    if (receiving) {
+        // someone is uploading a mission; reject another upload until
+        // they are done or timeout
+        send_mission_ack(_link, msg, MAV_MISSION_DENIED);
+        return;
+    }
+
+    if (packet.count > max_items()) {
+        send_mission_ack(_link, msg, MAV_MISSION_NO_SPACE);
+        return;
+    }
+
+    set_count(packet);
+
+    if (packet.count == 0) {
+        // no requests to send...
+        send_mission_ack(_link, msg, MAV_MISSION_ACCEPTED);
+        complete(_link);
+        return;
+    }
 
-    waypoint_dest_sysid = msg->sysid;       // record system id of GCS who wants to upload the mission
-    waypoint_dest_compid = msg->compid;     // record component id of GCS who wants to upload the mission
+    // start waypoint receiving
+    init_send_requests(_link, msg, 0, packet.count-1);
+}
+
+void MissionItemProtocol_Waypoints::set_count(const mavlink_mission_count_t &packet)
+{
+    // new mission arriving, truncate mission to be the same length
+    mission.truncate(packet.count);
 }
 
 /*
   handle a MISSION_CLEAR_ALL mavlink packet
  */
-void GCS_MAVLINK::handle_mission_clear_all(AP_Mission &mission, mavlink_message_t *msg)
+void GCS_MAVLINK::handle_mission_clear_all(const mavlink_message_t *msg)
 {
     // decode
     mavlink_mission_clear_all_t packet;
     mavlink_msg_mission_clear_all_decode(msg, &packet);
 
-    // clear all waypoints
-    if (mission.clear()) {
-        // send ack
-        mavlink_msg_mission_ack_send(chan, msg->sysid, msg->compid, MAV_MISSION_ACCEPTED,
-                                     MAV_MISSION_TYPE_MISSION);
-    }else{
-        // send nack
-        mavlink_msg_mission_ack_send(chan, msg->sysid, msg->compid, MAV_MISSION_ERROR,
-                                     MAV_MISSION_TYPE_MISSION);
+    const MAV_MISSION_TYPE mission_type = (MAV_MISSION_TYPE)packet.mission_type;
+    MissionItemProtocol *prot = gcs().get_prot_for_mission_type(mission_type);
+    if (prot == nullptr) {
+        send_mission_ack(*msg, mission_type, MAV_MISSION_UNSUPPORTED);
+        return;
     }
+
+    prot->handle_mission_clear_all(*this, *msg);
 }
 
-/*
-  handle a MISSION_WRITE_PARTIAL_LIST mavlink packet
- */
-void GCS_MAVLINK::handle_mission_write_partial_list(AP_Mission &mission, mavlink_message_t *msg)
+bool MissionItemProtocol_Waypoints::clear_all_items()
+{
+    return mission.clear();
+}
+
+bool MissionItemProtocol_Rally::clear_all_items()
+{
+    rally.set_rally_total(0);
```

Comment:
Scope creep, and I'm not going to hold you to it, but from an API perspective having AP_Rally have a `AP_Rally::clear()` method would make a bit more sense from the API sense. I know it has the same overall impact as what you've given here, but it won't necessarily in the future, and would allow someone like scripting to have a sane interface.

Reply:
`clear()` or `truncate(n)` like `AP_Mission`?

Reply:
I'm fine with `truncate(0)` `clear()` implies extra special stuff in that case, and `truncate(0)` would need to know to shell out to that anyways since it could happen.